### PR TITLE
Adding DeepCopy profiling hooks in deep_copy call

### PR DESCRIPTION
**Pull request by crtrott**
addresses issue #2522

I can't believe nobody notices this before ...

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
Yikes.

**Comment by DavidPoliakoff** [comment]
I'm moving "add testing for Kokkos Tools" to "absolute next task"

**Comment by crtrott** [comment]
Turns out somebody did notice, we just ignored Stan:  kokkos/kokkos-tools#49

**Comment by DavidPoliakoff** [comment]
Yeah, that was half of why I typed "yikes." I also need to do some triage on the repo, apparently

**Comment by crtrott** [comment]
@DavidPoliakoff are you reviewing? Take a good look at whether I added the "endDeepCopy" in all the necessary places (because there are early return statements in the code).

**Comment by DavidPoliakoff** [comment]
@crtrott, before I merge, I want to test what kinds of overheads we're adding. @dalg24 helpfully commented that we do this for rank-0 overloads, and I worry that turning essentially

```c++
memcpy(dst,src,sizeof(int))
```

into

```c++
// lots of tool code
memcpy(dst,src,sizeof(int))
// lots of tool code
```

Might cause us some problems. I'll try to have that done by end of day

File: core/src/Kokkos_CopyViews.hpp [hunk 1]
```diff
@@ -2531,16 +2629,29 @@ inline void deep_copy(
...
+  typedef typename dst_traits::memory_space dst_memory_space;
+  static_assert(std::is_same<typename dst_traits::non_const_value_type,
+                             typename dst_traits::value_type>::value,
+                "deep_copy requires non-const type");
+#if defined(KOKKOS_ENABLE_PROFILING)
+  if (Kokkos::Profiling::profileLibraryLoaded()) {
+    Kokkos::Profiling::beginDeepCopy(
+        Kokkos::Profiling::SpaceHandle(dst_memory_space().name()), dst.label(),
+        dst.data(), Kokkos::Profiling::SpaceHandle(Kokkos::HostSpace().name()),
+        "(none)", &value, sizeof(DT));
```
Comment (dalg24) [inline core/src/Kokkos_CopyViews.hpp]: I think I messed up here, I assumed the destination was rank-0

**Comment by DavidPoliakoff** [comment]
It turns out the performance hit isn't terrible. I think "deep_copy" of scalar->scalar is our worst case, and that only slows down 50% with the space_time tool. I'll do a review of the code, then approve

**Pull request merged**